File: leet_lint_code/leet_code/cpp/is_valid_bst.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 98.
// Date: 2019.7.8

#include <iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        
        bool isValidLeft = true;
        if (root->left != NULL) {
            isValidLeft = root->val > root->left->val && isValidBSTHelp(root->left, (numeric_limits<long>::min)(), root->val);
        }
        
        bool isValidRight = true;
        if (root->right != NULL) {
            isValidRight = root->val < root->right->val && isValidBSTHelp(root->right, root->val,(numeric_limits<long>::max)());
        }
        
        return isValidLeft && isValidRight;
    }
    
    bool isValidBSTHelp(TreeNode* root, long lower, long upper) {
        if (root == NULL) {
            return true;
        }   
        
        return root->val > lower && root->val < upper && 
            isValidBSTHelp(root->left, lower, root->val) &&
            isValidBSTHelp(root->right, root->val, upper);
    }  
};

File: leet_lint_code/leet_code/java/IsValidBST.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License
// Author: leetcode
// This is a answer of leet code problem 98.
// Date: 2019.7.7

public class IsValidBST {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }

        boolean leftIsValid = true;
        if (root.left != null) {
            leftIsValid = root.left.val < root.val && isValidBSTHelp(root.left, Long.MIN_VALUE, root.val);
        }

        boolean rightIsValid = true;
        if (root.right != null) {
            rightIsValid = root.right.val > root.val && isValidBSTHelp(root.right, root.val, Long.MAX_VALUE);
        }

        return leftIsValid && rightIsValid;
    }

    boolean isValidBSTHelp(TreeNode root, long lower, long upper) {
        if (root == null) {
            return true;
        }

        return root.val > lower && root.val < upper
                && isValidBSTHelp(root.left, lower, root.val) && isValidBSTHelp(root.right, root.val, upper);
    }
}
